src/main/java/com/walmartlabs/x12/asn856/DefaultAsn856TransactionSetParser.java

```diff
@@ -32,34 +34,155 @@
  *
  */
 public class DefaultAsn856TransactionSetParser extends AbstractTransactionSetParserChainable {
-    
+
     private static final Logger LOGGER = LoggerFactory.getLogger(DefaultAsn856TransactionSetParser.class);
-    
+
     public static final String ASN_TRANSACTION_TYPE = "856";
     public static final String ASN_TRANSACTION_HEADER = "BSN";
+    public static final String ASN_TRANSACTION_TOTALS = "CTT";
 
+    @Override
     protected boolean handlesTransactionSet(List<X12Segment> transactionSegments, X12Group x12Group) {
         // the first segment should be an ST with the 856 transaction type code
         return X12ParsingUtil.verifyTransactionSetType(transactionSegments, ASN_TRANSACTION_TYPE);
     }
 
+    /**
+     * it is assumed that this method is only called after getting true as a
+     * response from {@link handlesTransactionSet}
+     */
+    @Override
     protected X12TransactionSet doParse(List<X12Segment> transactionSegments, X12Group x12Group) {
         AsnTransactionSet asnTx = null;
 
         if (transactionSegments != null) {
-            // MUST be wrapped in a valid transaction set envelope
-            if (X12ParsingUtil.isValidEnvelope(transactionSegments, 
-                X12TransactionSet.TRANSACTION_SET_HEADER,
-                X12TransactionSet.TRANSACTION_SET_TRAILER)) {
+            asnTx = new AsnTransactionSet();
+            
+            this.doParsing(transactionSegments, asnTx);
+        }
+
+        return asnTx;
+    }
 
-                // TODO: make sure first segment is a BSN
+    protected void doParsing(List<X12Segment> transactionSegments, AsnTransactionSet asnTx) {
+        int segmentCount = transactionSegments.size();
 
-                asnTx = new AsnTransactionSet();
-                asnTx.setSampleAsnOnly("TEST");
-            }
+        int segementAfterHierarchicalLoops = this.findSegmentAfterHierarchicalLoops(transactionSegments);
+        List<X12Segment> loopSegments = transactionSegments.subList(2, segementAfterHierarchicalLoops);
```

Comment:
Is it safe to assume that the size of `transactionSegments` will be at least 3

Reply:
I added unit tests (and it wasn't)... good catch